### ISSUE TITLE
listKeys() executed right away with dependency specified

### ISSUE THREAD

**Issue by faceless7171** [Azure/azure-quickstart-templates#5831]
### Issue Details
In my case I have root template with external links to nested templates, I am trying to deploy storage account and then pass storage account key to the function app. So I am tryng to call listKeys(storageaccount_resourceId) in root template with dependsOn linked to this nested storage deployment. The actual behavior is that listKeys function is executed right away after start of the deployment (even when dependsOn specified), and failing with ResourceNotFound error.
I am not sure is that expected behavior or not, but it looks very strange that listKeys() is not waiting all dependencies.

You can see example template here:
https://raw.githubusercontent.com/faceless7171/arm-template-bug/master/bug.json

**Comment by djarvis** [comment]
Yep.  Same issue here.  In my case I am creating a CosmosDB account and a Web App that depends on it.  In the Web App settings I do a listkeys() on the CosmosDB account but it fails, saying that the resource does not exist.

Is there a more specific path I can use for DependsOn() so that my ARM will wait appropriately?

**Comment by navinmistry** [comment]
I got same issue where I'm deploying a functional with using existing storage account.
function app properties uses listKeys to setup few parameters as "WEBSITE_CONTENTAZUREFILECONNECTIONSTRING" and "AzureWebJobsStorage"

listKeys failing with "ResourceNotFound" error as soon as deployment begins

**Comment by djarvis** [comment]
I never came across any good answer to this so I just scripted it up into two templates, running one after the other.  The first deploys the database and the second deploys stuff that depends on it.

**Comment by APIWT** [comment]
What happens if you use the resourceId helper rather than “testbugstorage” inside depends on?

**Comment by PiotrPerak** [comment]
Can anyone tell if this is by design or a bug?

**Comment by JQUINONES82** [comment]
I have seen this issue.  If I run the deployment right after it works.  https://stackoverflow.com/questions/63083599/using-deployment-template-resource-and-receiving-resource-group-not-round-but-al

**Comment by JQUINONES82** [comment]
This does not work if the RG is created in the same template.  There is an issue where the keys can not be found during the pre deployment checks.  MS Supported recommended removing the RG deployment from the template and creating it first.  I just made a helper script to make the RG and then call the template.  It all worked afterwards.

**Comment by brianstringfellow** [comment]
@bmoore-msft I thought it might to helpful to others here to see your reply to the same or similar problem over in Stack Overflow. I have just now encountered the problem after refactoring my ARM templates for Service Fabric to make things more modular where I deploy my storage accounts by `Microsoft.Resources/deployments` and using `templateLink` and then try to `listKeys(...)` when creating the virtual machine scale sets. Tried `dependsOn` entries in various forms to no avail. https://stackoverflow.com/a/54636537/2778027

**Comment by bmoore-msft** [comment]
for the TLDR; version: https://bmoore-msft.blog/2020/07/26/resource-not-found-dependson-is-not-working/

**Comment by aaronhu922** [comment]
Any update of this issue? Dependson is not working with listKeys() referred to the deployment.
[resourceId('Microsoft.Resources/deployments','Storage')] is the resource depended on, but it reported right away not found when the deployment start.
```
{
      "type": "Microsoft.KeyVault/vaults/secrets",
      "apiVersion": "2019-09-01",
      "name": "[concat(parameters('keyVaultName'), '/', parameters('storageAccountName'), 'StorageConnectionString')]",
      "properties": {
        "value": "[concat('DefaultEndpointsProtocol=https;AccountName=', parameters('storageAccountName'), ';AccountKey=', listKeys(resourceId('Microsoft.Storage/storageAccounts', parameters('storageAccountName')), '2019-06-01').keys[0].value)]"
      },
      "dependsOn": [
        "[resourceId('Microsoft.Resources/deployments','Storage')]",
      ]
    },
```
